nomad/structs/diff.go

```diff
@@ -571,9 +571,118 @@ func (t *Task) Diff(other *Task, contextual bool) (*TaskDiff, error) {
 		diff.Objects = append(diff.Objects, altIDDiffs...)
 	}
 
+	// Actions diff
+	if aDiffs := actionDiffs(t.Actions, other.Actions, contextual); aDiffs != nil {
+		diff.Objects = append(diff.Objects, aDiffs...)
+	}
+
 	return diff, nil
 }
 
+func findActionMatch(action *Action, newActions []*Action, newActionMatches []int) int {
+	indexMatch := -1
+
+	for i, newAction := range newActions {
+		if newActionMatches[i] >= 0 {
+			continue
+		}
+
+		if action.Name == newAction.Name {
+			indexMatch = i
+			break
+		}
+	}
+
+	return indexMatch
+}
+
+func actionDiff(old, new *Action, contextual bool) *ObjectDiff {
+	diff := &ObjectDiff{Type: DiffTypeNone, Name: "Action"}
+	var oldPrimitiveFlat, newPrimitiveFlat map[string]string
+
+	if reflect.DeepEqual(old, new) {
+		return nil
+	} else if old == nil {
+		old = &Action{}
+		diff.Type = DiffTypeAdded
+		newPrimitiveFlat = flatmap.Flatten(new, nil, true)
+	} else if new == nil {
+		new = &Action{}
+		diff.Type = DiffTypeDeleted
+		oldPrimitiveFlat = flatmap.Flatten(old, nil, true)
+	} else {
+		diff.Type = DiffTypeEdited
+		oldPrimitiveFlat = flatmap.Flatten(old, nil, true)
+		newPrimitiveFlat = flatmap.Flatten(new, nil, true)
+	}
+
+	// Diff the primitive fields
+	diff.Fields = fieldDiffs(oldPrimitiveFlat, newPrimitiveFlat, contextual)
+
+	// Diff the Args field using stringSetDiff
+	if setDiff := stringSetDiff(old.Args, new.Args, "Args", contextual); setDiff != nil {
+		diff.Objects = append(diff.Objects, setDiff)
+	}
+
+	return diff
+}
+
+// actionDiffs diffs a set of actions. If contextual diff is enabled, unchanged
+// fields within objects nested in the actions will be returned.
+func actionDiffs(old, new []*Action, contextual bool) []*ObjectDiff {
+	// Handle trivial case.
+	if len(old) == 1 && len(new) == 1 {
+		if diff := actionDiff(old[0], new[0], contextual); diff != nil {
+			return []*ObjectDiff{diff}
+		}
+		return nil
+	}
+
+	// Initialize matching arrays with -1 to indicate unmatched Actions.
+	oldMatches := make([]int, len(old))
+	newMatches := make([]int, len(new))
+	for i := range oldMatches {
+		oldMatches[i] = -1
+	}
+	for i := range newMatches {
+		newMatches[i] = -1
+	}
```

Comment:
Is the idea here that we're allowing the order of action blocks to be switched around without it showing up as a plan update? I don't think we want to do that, because the scheduler will definitely pick this up as a change and then that'll be surprising for folks.

Reply:
I think this code is based on the service diff heuristic from https://github.com/hashicorp/nomad/pull/10965#issuecomment-914727717 where we were trying to fix an incorrect diff caused by the assumption that service names were unique. I'm pretty sure reordering blocks causes a diff, but it would be good to test 😅 

But by finding a "matching" block we can provide a better diff output when a change is detected.

I would need to look deeper into the code, but if `action.name` is expected to be unique then we don't need this heuristic.

Reply:
ah okay, this is my misinterpretation of what would be considered a change. I'll rework it.